File: agent/crates/fleet-client/src/stream.rs
use crate::types::{AgentEvent, ServerCommand};
use anyhow::Result;
use tokio::sync::mpsc;
use tonic::{transport::Channel, Request, client::Grpc, codec::ProstCodec, metadata::MetadataValue};
use tokio_stream::wrappers::ReceiverStream;

pub struct EventStreamManager;

impl EventStreamManager {
    pub async fn start(
        channel: Channel,
        token: String,
        events_rx: mpsc::Receiver<AgentEvent>,
    ) -> Result<mpsc::Receiver<ServerCommand>> {
        let mut client = Grpc::new(channel);
        let path = http::uri::PathAndQuery::from_static("/edr.fleet.FleetService/EventStream");
        
        let mut req = Request::new(ReceiverStream::new(events_rx));
        let meta_token = MetadataValue::try_from(format!("Bearer {}", token))?;
        req.metadata_mut().insert("authorization", meta_token);
        
        let response = client.streaming(req, path, ProstCodec::<AgentEvent, ServerCommand>::default()).await?;
        let mut stream = response.into_inner();
        
        let (server_cmd_tx, server_cmd_rx) = mpsc::channel(100);

        tokio::spawn(async move {
            while let Ok(Some(cmd)) = stream.message().await {
                if server_cmd_tx.send(cmd).await.is_err() {
                    break;
                }
            }
        });

        Ok(server_cmd_rx)
    }
}

File: agent/crates/fleet-client/src/stream.py
import asyncio
import logging

from fleet-client.types import AgentEvent, ServerCommand

logger = logging.getLogger(__name__)


class EventStreamManager:

    @staticmethod
    async def start(
        connection,
        token: str,
        events_queue: asyncio.Queue,
    ) -> asyncio.Queue:

        command_queue = asyncio.Queue(maxsize=100)

        async def event_generator():
            while True:
                event = await events_queue.get()
                yield event

        async def receive_commands():
            try:
                async for command in connection.open_stream(
                    event_generator(),
                    token,
                ):
                    await command_queue.put(command)

            except Exception as e:
                logger.warning(
                    "Event stream closed: %s",
                    e,
                )

        asyncio.create_task(receive_commands())

        return command_queue
